smartapps/ady624/webcore.src/webcore.groovy

```diff
@@ -2184,195 +1999,194 @@ public String mem(showBytes = true) {
-    	attributes: attributes(),
-        semaphore: semaphore,
-        semaphoreName: semaphoreName,
-        semaphoreDelay: semaphoreDelay,
+	def startTime = now()
+	semaphore = semaphore ?: 0
+	def semaphoreDelay = 0
+	def semaphoreName = semaphore ? "sph$semaphore" : ''
+	def waited = false
+	if (semaphore) {		
+		//if we need to wait for a semaphore, we do it here
+		def lastSemaphore
+		while (semaphore) {
+			lastSemaphore = lastSemaphore ?: (atomicState[semaphoreName] ?: 0)
+			if (!lastSemaphore || (now() - lastSemaphore > 10000)) {
+				semaphoreDelay = waited ? now() - startTime : 0
+				semaphore = now()
+				atomicState[semaphoreName] = semaphore
+				break
+			}
+			waited = true
+			pause(250)
+		}
+	}
+	def storageApp = !!fetchWrappers ? getStorageApp() : null
+	return [
+		enabled: !settings.disabled,
+		attributes: attributes(),
+		semaphore: semaphore,
+		semaphoreName: semaphoreName,
+		semaphoreDelay: semaphoreDelay,
 		commands: [
-        	physical: commands(),
+			physical: commands(),
 			virtual: virtualCommands(),
 			overrides: commandOverrides()
 		],
-        comparisons: comparisons(),
-        coreVersion: version(),
-    	contacts: [:],
-    	devices: (!!fetchWrappers ? (storageApp ? storageApp.listAvailableDevices(true) : listAvailableDevices(true)) : [:]),
-        virtualDevices: virtualDevices(),
-        globalVars: listAvailableVariables(),
-        globalStore: state.store ?: [:],
-        settings: state.settings ?: [:],
-        lifx: state.lifx ?: [:],
-        powerSource: state.powerSource ?: 'mains',
+		comparisons: comparisons(),
+		coreVersion: version(),
+		contacts: [:],
+		devices: (!!fetchWrappers ? (storageApp ? storageApp.listAvailableDevices(true) : listAvailableDevices(true)) : [:]),
+		virtualDevices: virtualDevices(),
+		globalVars: listAvailableVariables(),
+		globalStore: state.store ?: [:],
+		settings: state.settings ?: [:],
+		lifx: state.lifx ?: [:],
+		powerSource: state.powerSource ?: 'mains',
 		region: state.endpoint.contains('graph-eu') ? 'eu' : 'us',
-        instanceId: hashId(app.id),
-        sunTimes: getSunTimes(),
-        started: startTime,
-        ended: now(),
-        generatedIn: now() - startTime,
-        redirectContactBook: settings.redirectContactBook,
-        logPistonExecutions: settings.logPistonExecutions,
-        useLocalFuelStreams : useLocalFuelStreams(),
-        waitedAtSemaphore : waited
-    ] + (isHubitat() ? [        
-		hsmStatus: state.hsmStatus ?: location.hsmStatus,
-        colors: getColors()
-    ] : [:])
+		instanceId: hashId(app.id),
+		sunTimes: getSunTimes(),
+		started: startTime,
+		ended: now(),
+		generatedIn: now() - startTime,
+		redirectContactBook: settings.redirectContactBook,
+		logPistonExecutions: settings.logPistonExecutions,
+		useLocalFuelStreams : useLocalFuelStreams(),
+		waitedAtSemaphore : waited
+	] + (isHubitat() ? [	
+		hsmStatus: state.hsmStatus ?: location.hsmStatus //,
+//		colors: getColors()
```

Comment:
Colors was added to the rtData to prevent the child from having to spin up the parent multiple times. This has a performance impact.

Reply:
I found colors not commonly used, and it slowed all piston starts by having larger data passed when every piston starts.  So optimizing for the normal case when few colors are used.  The parent spins up on every piston start already.

Reply:
Agreed, in the past I saw that it was being called multiple times, but that's only when the piston editor is opened, and not on each run like I thought. Keep in mind though, that multiple calls to the parent from the child fire up a new parent state on each call to a parent method which has to hit the database.

Reply:
I read on each new event, there is a parent call for the event.   So this is every piston start (since it starts for events).   Are you seeing something different?  (handleEvents()) always initializes the data from the parent for each new event.  So the parent effectively handles all events...

Reply:
It does, but what I was saying before is that if you do 
parent.method1() and then call
parent.method1() again in the same piston

there is overhead to call the second parent method (loading of state, settings, saving of state). 

It doesn't matter in this case, though, since colorUtil isn't called every execution like I thought it was. Was just commenting on when you said it had to spin up anyway. In my example above it has to spin up twice.

Reply:
I see.  I think the constant transferring of constants between parent -> child needs to stop

I'll look at ways to do this to reduce the init data.  The colors is just one part of this.  There is a lot more.